src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter_BinaryOperator.cs

```diff
@@ -988,26 +986,40 @@ private BoundExpression MakeNullableHasValue(SyntaxNode syntax, BoundExpression
                 method: null,
                 constrainedToTypeOpt: null);
 
-            // (tempx.HasValue OP tempy.HasValue)
-            BoundExpression rightExpression = MakeBinaryOperator(
-                syntax: syntax,
-                operatorKind: rightOperator,
-                loweredLeft: callX_HasValue,
-                loweredRight: callY_HasValue,
-                type: boolType,
-                method: null,
-                constrainedToTypeOpt: null);
+            var canOmitHasValueCheck = yNonNull is { ConstantValue.IsDefaultValue: false } || xNonNull is { ConstantValue.IsDefaultValue: false };
 
             // result = (tempx.GetValueOrDefault() OP tempy.GetValueOrDefault()) &
             //          (tempx.HasValue OP tempy.HasValue)
-            BoundExpression binaryExpression = MakeBinaryOperator(
-                syntax: syntax,
-                operatorKind: BinaryOperatorKind.BoolAnd,
-                loweredLeft: leftExpression,
-                loweredRight: rightExpression,
-                type: boolType,
-                method: null,
-                constrainedToTypeOpt: null);
+            BoundExpression binaryExpression;
+            if (canOmitHasValueCheck)
```

Comment:
>canOmitHasValueCheck [](http://example.com/codeflow?start=16&length=20)

Is this logic correct for `<`, `<=`, `>`, `>=` operators? I think we need to add tests for all comparison operators that can come through here.